### PR TITLE
Revised training core

### PR DESCRIPTION
**Pull request by Drenderer**
### Key improvements:
- Split training ingredients into state (model and optimizer state) and static (optimizer, batcher, loss, etc.)
- Defined a new core `training_loop` function that only takes state, static, and callbacks as arguments. The loop updates the state.
- Removed `CallbackArgs` in favor of a `TrainingView`, which provides read-only access to static and read-write access to the state. This greatly simplifies the callbacks.
- Defined an ABC for Losses, which allows for custom gradient computations and implements model unwrapping by default (which avoids a redefinition of the loss in the fit function).
- Refactored datahandler.
- Removed the `HistoryCallback` in favor of a more general logging framework using a `MetricLogger` callback and a dict-like `History` object. This significantly improves the separation of concerns and enables the user to easily track custom metrics.
- Refactored `klax.fit` to recreate the old behaviour with the new components.

### CODE REVIEW

File: klax/_losses.py [hunk 1]
```diff
@@ -12,36 +12,40 @@
...
+    An instance of the loss class has two methods that are required for
+    [`klax.fit`][`fit`]: `value` and `value_and_grad`. In most cases the default
+    implementation should be used. `value` just [`klax.unwrap`][unwraps] the model
+    before computing the loss as specified in `__call__`, while `value_and_grad`
+    per default applies `jax.value_and_grad` to `value`. These functions can be
+    overwritten, for example to enable custom calculations of the gradients.
 
     Example:
         A simple custom loss that computes the mean squared error between
         the predicted values `y_pred` and true values `y` for in inputs `x` may
```
Comment (jaosch) [inline klax/_losses.py]: Typo: delete "in"

File: klax/_losses.py [hunk 2]
```diff
@@ -50,73 +54,112 @@ class Loss(Protocol):
...
+        batch_axes: PyTree[int | None, "T ..."],  # type: ignore
     ) -> Scalar:
         """Abstract method to compute the loss for a given model and data.
 
         Args:
             model: The model parameters or structure to evaluate the loss.
-            data: The input data or structure used for loss computation.
-            batch_axis: Specifies the axis or axes corresponding to the batch
+            batch: The input data or structure used for loss computation.
+            batch_axes: Specifies the axis or axes corresponding to the batch
```
Comment (jaosch) [inline klax/_losses.py:70]: I don't fully agree with this description, since batch_axes could also be, e.g., an `eqx.Module`.

File: klax/_losses.py [hunk 3]
```diff
@@ -50,73 +54,112 @@ class Loss(Protocol):
...
+
+        Returns:
+            Tuple of loss value and gradient with respect to the model.
 
+        """
+        return eqx.filter_value_and_grad(self.value)(model, batch, batch_axes)
 
-mse = MSE()
 
+def loss(func: Callable) -> Loss:
```
Comment (jaosch) [inline klax/_losses.py:160]: This decorator is very cool

File: klax/_training.py [hunk 4]
```diff
@@ -24,16 +25,57 @@
...
+        state.model = model
+        state.opt_state = opt_state
+        return state, value
+
+    for callback in callbacks:
+        callback.on_training_start(state, static)
+
+    for step in range(1, static.steps + 1):
+        state, step_loss = make_step(state, next(static.batcher))
+        if any(
```
Comment (jaosch) [inline klax/_training.py]: Insert the following comment above:

```
# Run all callbacks and break if any of them request termination of
# the training loop.
# Note! The square brackets are important. Otherwise the loop is
# terminated with the first callback that returns true. But we want
# to run all callbacks first and then decide, whether to terminate.
```

File: klax/_training.py [hunk 5]
```diff
@@ -24,16 +25,57 @@
...
     batch_data,
-    broadcast_and_get_size,
 )
 from ._losses import Loss, mse
-from ._wrappers import apply, unwrap
+from ._trainstate import TrainingState, TrainingStatic
+from ._wrappers import apply
+
+
+def training_loop(
```
Comment (jaosch) [inline klax/_training.py]: I suggest a name change to, e.g., `loop_training`, `loop_fit`, `fit_core`, `make_training_loop`, because a function should always start with a verb.

File: klax/_training.py [hunk 6]
```diff
@@ -140,64 +185,6 @@ def fit[T: eqx.Module, H: Callback](
         A tuple of the trained model and the loss history.
 
     """
-    # Braodcast the batch_axis to the data. While this happens again in the
-    # batch_data, doing it here allows the use of the broadcasted batch_axis in
-    # the loss function. If `batch_axis` is a prefix of `data`, this ensures
-    # that only leafs of type ArrayLike are vmapped. Thus it is possible to
-    # have data like `(str, array)` ans still use `batch_axis=0` instead of
-    # `batch_axis=(None, 0)`.
-    batch_axis, dataset_size = broadcast_and_get_size(data, batch_axis)
```
Comment (jaosch) [inline klax/_training.py:149]: Happy to see this go :)

File: klax/_training.py [hunk 7]
```diff
@@ -210,53 +197,42 @@ def make_step(batch, flat_model, optimizer, flat_opt_state):
     # initially
     model = apply(model)
```
Comment (jaosch) [inline klax/_training.py:248]: I agree that this should be called `rewarp` in the future.

File: klax/_training.py [hunk 8]
```diff
@@ -210,53 +197,42 @@ def make_step(batch, flat_model, optimizer, flat_opt_state):
...
+                partial(loss.value, batch_axes=batch_axes),
+            )
+        }
+        if validation_data is not None:
+            metric_defs["validation_loss"] = (
+                validation_data,
+                partial(loss.value, batch_axes=batch_axes),
+            )
+        history = HistoryCallback(
+            metric_defs=metric_defs,
```
Comment (jaosch) [inline klax/_training.py]: Does this `metric_def` mean that there is no lazy evaluation of the loss anymore?

File: klax/_callbacks.py [hunk 9]
```diff
@@ -211,57 +103,73 @@ def __repr__(self):
...
+        step_loss: Scalar,
+        static: TrainingStatic,
+    ):
+        """Record the losses and step count.
+
+        Called at each step during training.
+        """
+        if step % self.log_every == 0:
+            self.steps.append(self.step_offset + step)
+            self.metrics["step_loss"].append(step_loss)
```
Comment (jaosch) [inline klax/_callbacks.py]: I don't find the name `step_loss` very intuitive. I understand that there is one loss value for every recorded step, but calling it `step_loss` created more confusion than additional information in my opinion.

File: klax/_callbacks.py [hunk 10]
```diff
@@ -211,57 +103,73 @@ def __repr__(self):
...
+            for name, (data, metric_fn) in self.metric_defs.items():
+                metric_value = metric_fn(state.model, data)
```
Comment (jaosch) [inline klax/_callbacks.py]: How would we include a cache here?

I think this could be accoplished by somthing like creating a `TrainingCache` class that gets the loss functions from as inputs and instead of passing the loss functions directly to the the history, we passe the cached versions.

Or maybe we can create a `@cache_training` decorator, that can can be wrapped around any function and stores it results in a global cache.

**Review by jaosch — CHANGES_REQUESTED** [review]
Very cool changes. The main challenge will be to implement a cache, that is easy to reuse for any other callback.

File: klax/_callbacks.py (hunk 9 above)
Comment (Drenderer) [reply klax/_callbacks.py]: batch_loss maybe?

Comment (Drenderer) [reply klax/_callbacks.py]: Or minibatch_loss.

File: klax/_training.py (hunk 8 above)
Comment (Drenderer) [reply klax/_training.py]: Yes, and no. Klax no longer makes sure to cache loss itself. However, to make this cached just make sure that the metrics that you may compute multiple times each step are computed using a cached function, for example by decorating the function with @functools.cache.

File: klax/_callbacks.py (hunk 10 above)
Comment (Drenderer) [reply klax/_callbacks.py]: Just chache the metric_fn, with metric_fn = functools.cache(metric_fn), and also use the cached version elsewhere.

File: klax/_training.py (hunk 6 above)
Comment (Drenderer) [reply klax/_training.py:149]: Yeah! :D

**Comment by Drenderer** [comment]
> Very cool changes. The main challenge will be to implement a cache, that is easy to reuse for any other callback.

Thanks! :)
But I disagree about the caching, the user should just provide a cached function as metric_def.

File: klax/_training.py (hunk 7 above)
Comment (Drenderer) [reply klax/_training.py:248]: 👍

**Comment by Drenderer** [comment]
With the history callback now storing the metric definitions it is no longer possible to pickle it. This makes saving difficult. It would be possible to just not save the metric_defs, but that means that after loading the user needs to supply the defs again if they want to continue training. 😐

**Comment by jaosch** [comment]
> With the history callback now storing the metric definitions it is no longer possible to pickle it. This makes saving difficult. It would be possible to just not save the metric_defs, but that means that after loading the user needs to supply the defs again if they want to continue training. 😐

This is a similar problem to model serialization with hyperparameters (see [`https://docs.kidger.site/equinox/examples/serialisation/`]()). But I am not yet sure, what the solution could be.
